File: src/main/java/org/pyjjs/scheduler/core/actors/common/supervisor/SupervisorState.java
package org.pyjjs.scheduler.core.actors.common.supervisor;

import akka.actor.ActorContext;
import akka.actor.ActorRef;
import org.pyjjs.scheduler.core.actors.common.ActorState;
import org.pyjjs.scheduler.core.model.IdentifiableObject;

import java.util.HashMap;
import java.util.Map;

public abstract class SupervisorState<T extends IdentifiableObject> extends ActorState{

    private Map<T, ActorRef> actorsBySource = new HashMap<>();

    public SupervisorState(ActorContext actorContext) {
        super(actorContext);
    }

    public void registerTaskActor(T entity, ActorRef actorRef) {
        actorsBySource.put(entity, actorRef);
    }

    public void unregisterTaskActor(T entity) {
        actorsBySource.remove(entity);
    }

    public ActorRef findBySource(T entity) {
        return actorsBySource.get(entity);
    }

}

File: src/main/kotlin/org/pyjjs/scheduler/core/api/impl/actors/common/supervisor/SupervisorState.kt
package org.pyjjs.scheduler.core.api.impl.actors.common.supervisor

import akka.actor.ActorContext
import akka.actor.ActorRef
import org.pyjjs.scheduler.core.api.impl.actors.common.ActorState
import org.pyjjs.scheduler.core.model.IdentifiableObject

import java.util.HashMap

abstract class SupervisorState<in T : IdentifiableObject>(actorContext: ActorContext) : ActorState(actorContext) {

    private val actorsBySource = HashMap<T, ActorRef>()

    open fun registerTaskActor(entity: T, actorRef: ActorRef) {
        actorsBySource.put(entity, actorRef)
    }

    fun unregisterTaskActor(entity: T) {
        actorsBySource.remove(entity)
    }

    fun findBySource(entity: T): ActorRef? {
        return actorsBySource[entity]
    }

}
